ci: Custom GitHub Runners for Nix Builds

yvan-sraka opened the pull request:
This PR implements a migration from GitHub's standard runners to a hybrid infrastructure combining self-hosted and ephemeral Blacksmith runners for building Nix packages.
The implementation includes runner selection, dynamic build matrix generation, and optimized caching strategies to improve build performance and cost efficiency.

## Problem Statement

The previous CI implementation had several limitations:

1. **Monolithic build process**: A single job attempted to build all packages across all architectures
2. **Inefficient resource allocation**: All packages used the same runner type regardless of build complexity
3. **Limited parallelization**: Builds couldn't be efficiently distributed across different runner types
4. **Redundant builds**: No mechanism to skip packages already available in the binary cache
5. **Poor cost optimization**: Large, expensive builds ran on the same infrastructure as small, quick builds
6. **Poor job output clarity**: No separation of build results made it hard to identify issues

## Solution Architecture

### High-Level Design

```
┌─────────────────┐
│   nix-eval      │  Evaluates flake, generates build matrix
│   (Blacksmith)  │  Identifies cached vs. uncached packages
└────────┬────────┘  Identifies large packages
         │
         ├──────────────┬──────────────┬
         │              │              │              
         v              v              v              
┌────────────────┐ ┌────────────────┐ ┌────────────────┐
│ aarch64-linux  │ │ aarch64-darwin │ │ x86_64-linux   │
│ Self-hosted/   │ │ Self-hosted    │ │ Blacksmith     │
│ Blacksmith     │ │ (macOS)        │ │ Ephemeral      │
└────────────────┘ └────────────────┘ └────────────────┘
```

### Architecture Components

1. **Nix Evaluation Phase** (`nix-eval.yml`):
   - Runs on powerful ephemeral runner (32vcpu)
   - Evaluates all flake outputs using `nix-eval-jobs`
   - Checks cache status for each package
   - Generates optimized build matrices per architecture

2. **Build Phases** (separate jobs per architecture):
   - **aarch64-linux**: Self-hosted or Blacksmith ARM runners
   - **aarch64-darwin**: Self-hosted macOS runners
   - **x86_64-linux**: Blacksmith ephemeral runners

3. **Runner Selection Logic**:
   - KVM-required packages → Self-hosted runners with KVM support
   - Large packages (Rust, PostGIS) → 32vcpu runners
   - Standard packages → 8vcpu runners
   - Darwin packages → Self-hosted macOS runners

## Key Components

### 1. Dynamic Matrix Generation (`github-matrix` Package)

**Location**: `nix/packages/github-matrix/`

**Core Responsibilities**:
- Evaluates Nix flake outputs using `nix-eval-jobs` (https://github.com/nix-community/nix-eval-jobs)
- Determines package dependencies and build order using topological sorting
- Identifies cached packages to skip redundant builds
- Assigns appropriate runners based on package requirements
- Generates GitHub Actions-compatible JSON matrices

**Package Size Detection**:
- Uses `requiredSystemFeatures = ["big-parallel"]` in package definitions
- Automatically allocates 32vcpu runners for:
  - Rust-based extensions (pg_graphql, pg_jsonschema, wrappers)
  - PostGIS (complex C++ builds)
  - pgvector with heavy dependencies

**Output Format**:
```json
{
  "aarch64_linux": {
    "include": [
      {
        "attr": "checks.aarch64-linux.pg_graphql_15",
        "name": "pg_graphql-15.7",
        "system": "aarch64-linux",
        "runs_on": {"labels": ["blacksmith-32vcpu-ubuntu-2404-arm"]},
        "postgresql_version": "15"
      }
    ]
  },
  "x86_64_linux": {...},
  "aarch64_darwin": {...}
}
```

### 2. Custom Nix Installation Actions

Unify Nix installation across different runner types with two reusable GitHub Actions.

#### Ephemeral Runners (`nix-install-ephemeral`)

**Location**: `.github/actions/nix-install-ephemeral/`

**Purpose**: Set up Nix on fresh Blacksmith runners where Nix is not pre-installed

**Features**:
- Installs Nix 2.31.2 using cachix/install-nix-action
- Configures binary cache substituters
- Optionally sets up AWS credentials for cache pushing
- Creates post-build hook for automatic cache uploads

**Configuration**:
```yaml
- uses: ./.github/actions/nix-install-ephemeral
  with:
    push-to-cache: 'true'  # Enable for build jobs
  env:
    DEV_AWS_ROLE: ${{ secrets.DEV_AWS_ROLE }}
    NIX_SIGN_SECRET_KEY: ${{ secrets.NIX_SIGN_SECRET_KEY }}
```

**Cache Upload Mechanism**:
- Post-build hook automatically uploads successful builds to S3
- Uses Nix signing keys for trusted binary cache
- Hook script: `/etc/nix/upload-to-cache.sh`

#### Self-Hosted Runners (`nix-install-self-hosted`)

**Location**: `.github/actions/nix-install-self-hosted/`

**Purpose**: Configure AWS credentials on persistent self-hosted runners where Nix is pre-installed

**Features**:
- Assumes AWS IAM role via OIDC
- Writes credentials to `/etc/nix/aws/nix-aws-credentials`
- Supports custom role duration (default 5 hours)

### 3. Reusable Nix Eval Workflow

**Location**: `.github/workflows/nix-eval.yml`

**Purpose**: Shared workflow for matrix generation

**Features**:
- Callable from other workflows via `workflow_call`
- Outputs structured JSON matrix
- Runs on high-performance ephemeral runner
- Handles optional AWS credentials for cache access

### 4. Restructured Build Workflow

**Location**: `.github/workflows/nix-build.yml`

**New Structure**:

```yaml
jobs:
  nix-eval:
    # Generate build matrices
    uses: ./.github/workflows/nix-eval.yml

  nix-build-aarch64-linux:
    needs: nix-eval
    strategy:
      matrix: ${{ fromJSON(needs.nix-eval.outputs.matrix).aarch64_linux }}
    # Build ARM Linux packages

  nix-build-aarch64-darwin:
    needs: nix-eval
    strategy:
      matrix: ${{ fromJSON(needs.nix-eval.outputs.matrix).aarch64_darwin }}
    # Build macOS ARM packages

  nix-build-x86_64-linux:
    needs: nix-eval
    strategy:
      matrix: ${{ fromJSON(needs.nix-eval.outputs.matrix).x86_64_linux }}
    # Build x86_64 Linux packages

  run-testinfra:
    needs: [nix-build-aarch64-linux, ...]
    # Only run if all builds succeed or skip

  run-tests:
    needs: [nix-build-aarch64-linux, ...]
    # Run test suite
```

**Key Improvements**:

1. **Parallel Architecture Builds**: Each architecture builds independently
2. **Smart Job Skipping**: Uses `!cancelled()` with success/skip conditions
3. **Dynamic Job Names**: Include PostgreSQL version for clarity

## Related PRs

- **PR #1742**: Base work for custom runners (this PR builds on top of it)

yvan-sraka committed:
chores(ci): opted-out nix-fast-build

yvan-sraka committed:
chores(ci): set up nix-github-actions

yvan-sraka committed:
chores(ci): comment out vestigal testinfra-ami-build

yvan-sraka committed:
chores(ci): use custom github runners

samrose commented on .github/workflows/nix-build.yml:
```diff
@@ -104,23 +113,12 @@ jobs:
           AWS_SECRET_ACCESS_KEY: ${{ env.AWS_SECRET_ACCESS_KEY }}
           AWS_SESSION_TOKEN: ${{ env.AWS_SESSION_TOKEN }}
 
-  run-testinfra:
```
I think we jut need to add these tests back in

yvan-sraka committed:
chores(ci): use nix-eval-jobs and sets AWS creds to /etc/nix/aws

jfroche committed:
feat(ci): do not build if already cached

jfroche committed:
chore: improve reproducibility of postgresql builds

Do not create postgresql package that depends on the runtime evaluation context as
that harms reproducibility.

jfroche committed:
chore: remove nix-github-actions dependency

jfroche committed:
feat(ci): split nix build workflow into separate extensions and checks jobs

Split the monolithic nix-build job into two workflows:
one for building PostgreSQL extensions and another for checks.

Building extensions can be resource-intensive and time-consuming, 
so isolating them allows for better resource allocation and parallelism.

Once they are built, the checks job can run tests and validations
on the already built extensions.

jfroche committed:
feat(ci): extract nix build setup into reusable action and split builds by architecture

Extract AWS credential setup and nix build steps into a composite action
to reduce duplication. Split extension builds into separate jobs per
architecture (aarch64-linux, aarch64-darwin, x86_64-linux) and update
matrix generation to group packages by system.

jfroche committed:
fix(ci): do not hide cached builds

jfroche committed:
Revert "fix(ci): do not hide cached builds"

This reverts commit 9d41b1e45fd92df646baaf078b9889cf308ac32b.

jfroche committed:
chore: Temporarily disable x86_64-linux builds

We don't have a self hosted runner for x86_64-linux at the moment

jfroche committed:
fix: sort packages and filter out cached ones

jfroche committed:
fix: do not skip checks-matrix if dependencies are skipped

jfroche committed:
fix: do not return debug fields in GitHub Actions matrix output

jfroche committed:
debugging

jfroche committed:
fix(ci): use !cancelled() instead of always() for dependent job conditions

Replace always() with !cancelled() to prevent jobs from running when the workflow
is cancelled. Add explicit success/skip conditions for dependent jobs to ensure
proper workflow execution flow.

jfroche committed:
fix(ci): stop chaining aws roles

We couldn't set duration to 18000 seconds when chaining roles.

jfroche committed:
fix(ci): disable eval-cache and accept-flake-config

jfroche committed:
ci: run nixos test on aarch64-linux

jfroche committed:
ci: split checks build jobs by system architecture

Refactor GitHub Actions workflow to run build checks in parallel across different
architectures (aarch64-linux, aarch64-darwin) with separate job matrices.

jfroche committed:
fix(ci): do not try to build already cached checks

jfroche committed:
fix(ci): use correct architecture name in aarch64-linux builds

jfroche committed:
fix(ci): simplify GitHub Actions workflow for Nix builds

Create a single nix-eval job to determine packages to build, removing
redundant extension and check matrices.

jfroche committed:
Revert "chores(ci): comment out vestigal testinfra-ami-build"

This reverts commit e2db36800b8a2a0ad9e132fdad688613aeb5aa37.

jfroche committed:
feat(ci): eval on blacksmith-32vcpu-ubuntu-2404

jfroche committed:
feat: add ephemeral Nix install action for GitHub runners

jfroche committed:
refactor(ci): extract nix eval into reusable workflow

jfroche committed:
feat: enable x86_64-linux builds in CI

jfroche committed:
feat: add PostgreSQL version to GitHub Actions job names

When building a postgres extension, the build matrix may include
multiple time the same extension for different PostgreSQL versions.
This change makes it easier to identify which job corresponds to which PostgreSQL
version in the workflow runs.

jfroche committed:
feat: run actionlint on new GitHub Actions workflows

jfroche committed:
fix: disable treefmt flake check

treefmt is already included in the pre-commit hooks check.

jfroche committed:
chore: improve github matrix script type annotations

jfroche committed:
feat: optimize CI runner selection based on package size

Dynamically assign larger runners (32vcpu) for Rust and PostGIS extensions
while using smaller runners (8vcpu) for standard packages.

jfroche committed:
chore: fix package meta maintainers format

jfroche committed:
chore: create a nix package for generating GitHub Actions matrix

Add pytest tests for the package
Add nix-eval-jobs in path for the package

jfroche committed:
fix: configure runner according to the matrix job

The matrix job returns the type of runner, so we can configure the nix
installation step accordingly.

jfroche committed:
Update nix-eval-jobs

Our changes were merged upstream, so we can now track the original
repository again.

jfroche committed:
refactor(ci): standardize nix installation and disable cache push by default

- Replace DeterminateSystems/nix-installer-action with custom nix-install-ephemeral action across all workflows
- Change default push-to-cache from 'true' to 'false' to prevent unnecessary nix/aws configurations
- Explicitly enable push-to-cache only for nix-build and nix-eval workflows where caching is beneficial

jfroche committed:
fix: nix devShell inclusion condition

jfroche committed:
fix(ci): ensure x86_64-linux build is considered in testinfra and test workflows

jfroche committed:
feat: use big-parallel to identify large packages

jfroche committed:
fix(ci): eval should fail if github-matrix run fails

yvan-sraka committed:
fix(ci): remove redundant build psql bundle step

jfroche committed:
fix: reduce ARM runner size from 8vcpu to 4vcpu for ephemeral builds

We might not need the full 8vcpu for aarch64-linux builds, so this
change reduces the runner size to 4vcpu to wait less for available
blacksmith runners.

samrose reviewed (changes requested):
I don’t see any reason we can’t merge reviewing code. Just need to generate images and test

samrose commented:
Added request change just to block merge until we test

samrose commented:
Noting will need to tell folks to enable big-parallel on local build of rust based packages that require that (I am assuming) and will need to start aligning people to use the same version of nix we use here.

samrose reviewed (changes requested):
just adding this to prevent merge until we know this is ready

jfroche committed:
feat: fail pipeline if nix evaluation fails

jfroche committed:
feat: do not return empty matrices if no package has to be built

samrose commented on nix/ext/pgvector.nix:
```diff
@@ -69,7 +69,7 @@ let
         description = "Open-source vector similarity search for Postgres";
         homepage = "https://github.com/${src.owner}/${src.repo}";
-        maintainers = with maintainers; [ olirice ];
+        maintainers = [ "olirice" ];
```
```suggestion

```
fwiw we can just remove this like since we decided to drop naming "maintainers" in supabase/postgres packages long ago

samrose commented:
overall looks good. Want to test workflow_dispatch of ami release to be sure everything is working on that level.

yvan-sraka committed:
Update nix/ext/pgvector.nix

Co-authored-by: samrose <samuel@supabase.io>

samrose commented:
tested ami creation and functionality of project instance, upgrades, pause/restore

jfroche committed:
fix: add skip job only for systems that don't have any job

jfroche committed:
fix(github-matrix): handle evaluation errors without deadlock

Fix github-matrix that would hang when nix-eval-jobs encountered errors due to subprocess pipe deadlock - stderr buffer would fill while reading stdout.

This change ensure that evaluation errors are visible and the workflow fails properly while still showing which packages succeeded.

jfroche committed:
feat(github-matrix): integrate github-action-utils for better error visibility

Integrates github-action-utils library to improve error and warning
visibility in GitHub Actions UI through workflow command annotations.

jfroche committed:
feat(github-matrix): group evaluation errors by message

Refactor error handling to collect and group evaluation errors similar to warnings. Errors with the same message are now displayed together with a list of affected attributes.

jfroche committed:
fix(github-matrix): improve multiline error display in GitHub Actions

Extract core error messages and format them better for GitHub Actions
annotations.

jfroche committed:
fix(ci): skip run-testinfra and run-tests when nix-eval fails

Add nix-eval to needs dependencies and check its result in conditional expressions to prevent downstream test jobs from running when evaluation fails.

jfroche committed:
chore(github-matrix): update message when there are no build for a system

jfroche committed:
fix(github-matrix): backward compatibility for Result access

We are running an older version of the 'result' library that uses
'_value' instead of 'ok_value' to access the successful result of a
computation.

jfroche committed:
refactor: migrate from packages to legacyPackages for PostgreSQL extensions

To be able to build extensions versions packages separately in CI, we
need to expose them in a nested structure. It is currently not possible
to do so with the flattened packages structure, as the individual
extension packages are not directly accessible.

In this change, we replace the flattened package structure with nested
legacyPackages to improve discoverability of individual extension
packages.

jfroche committed:
refactor(nix): remove "-all" suffix from extension package names

Simplify extension package naming by removing the redundant "-all" suffix that was appended to pname attributes.

jfroche committed:
refactor(ci): split nix build jobs into separate packages and checks workflows

To make sure we only build what is necessary, we start building packages first, 
then run checks once all packages are built successfully.

jfroche committed:
feat: use 8 vCPU runner for aarch64 builds

Use the same 8 vCPU runner for aarch64 builds as used for x86_64 builds to improve build performance.

jfroche committed:
feat(ci): test blacksmith stickydisk for eval
